### PR TITLE
Fix casting issue with immutable-typed fields of record type

### LINKED ISSUE #30025: Bad, sad from use of readonly
See attached.

This goes away if `ops` is declared as `final BasicTypeOps[] ops`.

[semtype.zip](https://github.com/ballerina-platform/ballerina-lang/files/6325262/semtype.zip)

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #30025 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java [hunk 1]
```diff
@@ -504,16 +503,17 @@ private static void populateImmutableStructureFields(Types types, SymbolTable sy
                                                         unresolvedTypes);
 
             Name origFieldName = origField.name;
-            BVarSymbol immutableFieldSymbol = new BVarSymbol(origField.symbol.flags | flag,
-                                                             origFieldName, pkgID, immutableFieldType,
-                                                             immutableStructureSymbol, origField.pos, SOURCE);
+            BInvokableSymbol immutableFieldSymbol = new BInvokableSymbol(origField.symbol.tag,
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/util/ImmutableTypeCloner.java:506]: Is this the correct fix? I don't think using BInvokableSymbol is correct when the type of the field is not a function type.

You can check with something like
```ballerina
type Foo record {|
    int a;
    function () b;
    int|function () c;
|};
```

**Review by MaryamZi — COMMENTED** [review]
Shall we also fix the following?

```ballerina
type Foo record {|
    function () x?;
|};

public function main() {
    record {| function () x; int i?; |} rec = {x: func};

    if rec is Foo {

    }
}

function func() {

}
```

We seem to be using similar code there, and it crashes on master with the same CCE.

File: tests/jballerina-unit-test/src/test/resources/test-src/record/readonly_record_fields.bal [hunk 2]
```diff
@@ -171,6 +171,19 @@ type Identifier record {|
...
 |};
 
+type Foo2 record {|
+    function () f;
+    int id;    
+|};
+
+function testRecordWithFunctionTypeField() {
+    Foo2 & readonly x = {
+        f: testRecordWithFunctionTypeField,
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/record/readonly_record_fields.bal:181]: Let's add some tests where we call these functions too. Let's use different combinations of param and return types.

**Comment by chiranSachintha** [comment]
> Shall we also fix the following?
> 
> ```ballerina
> type Foo record {|
>     function () x?;
> |};
> 
> public function main() {
>     record {| function () x; int i?; |} rec = {x: func};
> 
>     if rec is Foo {
> 
>     }
> }
> 
> function func() {
> 
> }
> ```
> 
> We seem to be using similar code there, and it crashes on master with the same CCE.

Fixed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java [hunk 3]
```diff
@@ -4227,17 +4227,21 @@ private boolean populateFields(IntersectionContext intersectionContext, BRecordT
             }
 
             org.wso2.ballerinalang.compiler.util.Name name = lhsRecordField.name;
-            BVarSymbol recordFieldSymbol = new BVarSymbol(intersectionFlags, name, env.enclPkg.packageID,
-                                                          intersectionFieldType, newTypeSymbol, lhsRecordField.pos,
-                                                          SOURCE);
+            BVarSymbol recordFieldSymbol;
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4230]: Please add a test for this too.

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java:4230]: Added.

**Review by MaryamZi — APPROVED** [review]

**Pull request merged**